qubes/vm/mix/dvmtemplate.py

```diff
@@ -45,14 +45,12 @@ def __on_pre_del_dvmtemplate(self, event, name,
         self.__on_pre_set_dvmtemplate(
             event, name, False, oldvalue)
 
-    @qubes.events.handler('property-pre-set:template')
-    def __on_property_pre_set_template(self, event, name, newvalue,
+    @qubes.events.handler('property-set:template')
+    def __on_property_set_template(self, event, name, newvalue,
             oldvalue=None):
         # pylint: disable=unused-argument
-        if any(self.dispvms):
-            raise qubes.exc.QubesVMInUseError(self,
-                'Cannot change template '
-                'while there are DispVMs based on this qube')
+        for vm in self.dispvms:
+            vm.on_property_set_template(event, name, newvalue, oldvalue)
```

Comment:
I think you need also a pre-set handler to check if there is no running DispVM. Perhaps DispVMs with `auto_cleanup=True` could be excluded from this check (as there isn't a risk of starting it again with not switched template), but not fully sure about that - qvm-prefs will tell lies for such not updated DispVMs.

In the end, we want to have some queued template change feature: https://github.com/QubesOS/qubes-issues/issues/5697

Reply:
What happens if there is no pre-set handler, and `on_property_set_template` throws?

Reply:
Property remains set, other handlers may not be executed.